exonum/src/blockchain/transaction.rs

```diff
@@ -0,0 +1,398 @@
+    ///         // ...
+    ///
+    ///         // Return execution status.
+    ///         Ok(())
+    ///     }
+    ///
+    ///     // Other methods...
+    ///     // ...
+    /// #   fn verify(&self) -> bool { true }
+    /// }
+    /// # fn main() {}
+    fn execute(&self, fork: &mut Fork) -> ExecutionResult;
+}
+
+/// Result of unsuccessful transaction execution.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum ExecutionError {
+    /// General failure (unspecified reason).
+    Failure,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+    /// User-defined string error description.
+    Description(String),
+}
+
+/// Extended by the framework result of unsuccessful transaction execution.
+///
+/// # Notes:
+///
+/// - `Description`'s content excluded from hash calculation (see `StorageValue` implementation for
+///   the details).
+/// - `TransactionError::Panic` is set by the framework if panic is raised during transaction
+///   execution.
+/// - `TransactionError` implements `Display` which can be used for obtaining a simple error
+///   description.
+///
+/// # Examples
+///
+/// ```
+/// use exonum::blockchain::TransactionError;
+///
+/// # let transaction_error = TransactionError::Panic;
+/// // Prints user friendly error description.
+/// println!("Transaction error: {}", transaction_error);
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub enum TransactionError {
+    /// Panic occurred during transaction execution.
+    Panic,
+    /// General failure (unspecified reason).
+    Failure,
+    /// User-defined error-code. Can have different meanings for different transactions and
+    /// services.
+    Code(u8),
+    /// User-defined string error description.
+    Description(String),
+}
+
+/// Converts from `ExecutionResult` into `TransactionResult`.
+pub fn convert_result(result: ExecutionResult) -> TransactionResult {
+    match result {
+        Ok(()) => Ok(()),
+        Err(ExecutionError::Failure) => Err(TransactionError::Failure),
+        Err(ExecutionError::Code(c)) => Err(TransactionError::Code(c)),
+        Err(ExecutionError::Description(s)) => Err(TransactionError::Description(s)),
+    }
+}
+
+impl<'a, T: Transaction> From<T> for Box<Transaction + 'a> {
+    fn from(tx: T) -> Self {
+        Box::new(tx) as Box<Transaction>
+    }
+}
+
+impl fmt::Display for TransactionError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            TransactionError::Panic => write!(f, "Panic during execution"),
+            TransactionError::Failure => write!(f, "Unspecified failure"),
+            TransactionError::Code(c) => write!(f, "Error code: {}", c),
+            TransactionError::Description(ref s) => write!(f, "{}", s),
+        }
+    }
+}
+
+// String content (`TransactionError::Description`) is intentionally excluded from the hash
```

Comment:
I don't find this reasoning particularly compelling. Consider that excluding a message from the hash calculation means that it could be faked for lightweight clients. Moreover, I'd say that doing this *encourages* users to use error descriptions as they won't break the blockchain.

Maybe, do the following: make a string description an optional field in `ExecutionError` (remaking it into a struct with status and this message fields). Then the message would not be committed into the hash by design, but a service dev would need to specify a status together with it to indicate what's going on and to commit *something* to the blockchain state.

Reply:
I understand your concern, but I'm not sure that the proposed solution will make a big difference: user can still use different descriptions with the same code. Am I missing something?

Reply:
Discussed: I'm convinced now.